### PR TITLE
CONTRIBUTING.md: Add PR closing policy

### LINKED ISSUE #11583: Policy for stale PRs
We currently have 71 open PRs. That's not terrible (mypy and CPython have a lot more), but it's more than I'd like. Many of the PRs are not in a mergeable state, so it's hard to get the number down.

I'd like to propose a few policies to help keep the number under control:

- "Deferred" PRs (https://github.com/python/typeshed/pulls?q=is%3Aopen+is%3Apr+label%3Adeferred) must have a clearly stated blocker. That can be either an unambiguous bug in a type checker (the type checker is not following the spec) or a dependency on an active PEP.
- It is the responsibility of the PR author to ensure that all of our tests pass. If a PR has been open for *three months* and CI is still failing, we can close it. This includes draft PRs.

If other maintainers agree, I can add these guidelines to our CONTRIBUTING documentation.

### PR DESCRIPTION
**Pull request by JelleZijlstra**
Closes #11583

### CODE REVIEW

**Review by srittau — APPROVED** [review]
Looks great!

File: CONTRIBUTING.md
Comment (AlexWaygood) [inline CONTRIBUTING.md:704]: I don't want to commit us to doing this for _anything_ that might be blocked

```suggestion
We sometimes use the ["deferred" label](https://github.com/python/typeshed/labels/deferred)
```

File: CONTRIBUTING.md
Comment (AlexWaygood) [inline CONTRIBUTING.md:715]:
```suggestion
the blocking issue resolved within a reasonable time frame. If a PR depends on
a more amorphous change, such as a type system change that has not yet reached
the PEP stage, it should instead be closed.
```

File: CONTRIBUTING.md
Comment (AlexWaygood) [inline CONTRIBUTING.md:724]:
```suggestion
open for too long. Specifically, we close open PRs that either have failures in CI,
serious merge conflicts or unaddressed feedback, and that have not seen any
activity in three months.
```

File: CONTRIBUTING.md
Comment (AlexWaygood) [inline CONTRIBUTING.md:732]: Do contributors have the ability to reopen PRs that have been closed by maintainers? I'm not sure

**Review by AlexWaygood — COMMENTED** [review]
Thanks!

File: CONTRIBUTING.md
Comment (AlexWaygood) [reply CONTRIBUTING.md:732]: Here you go:
- https://github.com/srittau/python-asserts/pull/101

Comment (AlexWaygood) [reply CONTRIBUTING.md:732]: Seems like it is indeed impossible for contributors without write access to reopen a PR if a maintainer has closed it: https://github.com/srittau/python-asserts/pull/101#issuecomment-1997618854

```suggestion
    Therefore, I am closing this PR. If you are still interested, please feel
    free to open a new PR (or ping us to reopen this one)
```

File: CONTRIBUTING.md
Comment (JelleZijlstra) [inline CONTRIBUTING.md:734]:
```suggestion
    Thanks for your contribution to typeshed! Unfortunately, this PR still
    (fails some tests OR has unresolved review feedback OR has a merge conflict) after three months of inactivity.
    Therefore, I am closing this PR. If you are still interested, please feel
    free to open a new PR (or ping us to reopen this one)
```
@Akuli how is this? Sorry your previous suggestion no longer applies because I accepted another change.

Comment (Akuli) [reply CONTRIBUTING.md:734]: It is better, but the `Thanks ...! Unfortunately ... Therefore ...` structure still feels a bit too formal to me. Maybe it's just me.

Comment (Akuli) [reply CONTRIBUTING.md:734]: Maybe:
```suggestion
    Thanks for contributing! I'm closing this PR for now, because it still
    (fails some tests OR has unresolved review feedback OR has a merge conflict) after three months of inactivity.
    If you are still interested, please feel
    free to open a new PR (or ping us to reopen this one)
```

Comment (AlexWaygood) [reply CONTRIBUTING.md:734]: Akuli's suggestion looks like a nice improvement to me too (though I'm also okay with the current wording)

Comment (srittau) [reply CONTRIBUTING.md:734]: Missing period at the end of the sentence, but sounds good to me. I like that you have to choose the exact reason.

Comment (Akuli) [reply CONTRIBUTING.md:734]:
```suggestion
Maybe:
```suggestion
    Thanks for contributing! I'm closing this PR for now, because it still
    (fails some tests OR has unresolved review feedback OR has a merge conflict) after three months of inactivity.
    If you are still interested, please feel free to open a new PR (or ping us to reopen this one).
``````

File: CONTRIBUTING.md
Comment (Akuli) [inline CONTRIBUTING.md:733]: Nit: Let's wrap long lines.
```suggestion
    (fails some tests OR has unresolved review feedback OR has a merge conflict)
    after three months of inactivity. If you are still interested, please feel free to open
    a new PR (or ping us to reopen this one).
```

**Comment by Akuli** [comment]
(My last commit only wraps long lines. It doesn't change the wording.)

It seems like we're happy with this, but we could leave it open for a couple days or so in case other maintainers have something to say.

File: CONTRIBUTING.md
Comment (Avasam) [inline CONTRIBUTING.md:636]:
```suggestion
([see below](#closing-stale-prs)).
```

File: CONTRIBUTING.md
Comment (Avasam) [inline CONTRIBUTING.md:711]: This reminds me of the [implementation-pending](https://github.com/python/typeshed/labels/implementation-pending) label. How should that interact with this policy and the "deferred" label?

File: CONTRIBUTING.md
Comment (Avasam) [inline CONTRIBUTING.md:734]: Nitpick: one needs to be filled in, but the other is part of the comment. Maybe something like this?
```suggestion
    <fails some tests OR has unresolved review feedback OR has a merge conflict>
    after three months of inactivity. If you are still interested, please feel free to open
    a new PR (or ping us to reopen this one).
```

Also TIL you can make a text code block in md just by using indents

File: CONTRIBUTING.md
Comment (AlexWaygood) [reply CONTRIBUTING.md:711]: probably we don't really need both labels -- we could probably merge them into one :)

File: CONTRIBUTING.md
Comment (Avasam) [reply CONTRIBUTING.md]: > (My last commit only wraps long lines. It doesn't change the wording.)

It does, however, change the line-wrapping by adding forced word-break locations. If this is a block of text that we intend on copy-pasting, may has well save the maintainer from having to fix word wrapping, independent of line-length in the .md file.

Not a blocker, just a nitpick/opinion that this may be a special case.

File: CONTRIBUTING.md
Comment (srittau) [reply CONTRIBUTING.md:711]: We have too many labels anyway. They could use a good cleanup. But I agree that merging implementation-pending into deferred seems like an easy first step.

Comment (AlexWaygood) [reply CONTRIBUTING.md:711]: the `project` and `project-discussion` labels could also probably be merged

Comment (JelleZijlstra) [reply CONTRIBUTING.md:711]: I guess the idea of "implementation-pending" is that we're specifically waiting for a third-party library we're stubbing to release their implementation. That's maybe a little different from "deferred", but not sure it's worth a separate label.

**Review by Avasam — APPROVED** [review]

**Review by Akuli — APPROVED** [review]

**Pull request merged**
